lib/Partitioner/Partitioner.cpp

```diff
@@ -0,0 +1,324 @@
+  bfs.levels.push_back({level, {node}});
+  bfs.visited.insert(node);
+  level++;
+  while (current < level) {
+    std::vector<Node *> nList;
+    for (int i = 0, e = bfs.levels[current].second.size(); i < e; i++) {
+      Node *N = bfs.levels[current].second[i];
+
+      for (int j = 0, e = N->getNumInputs(); j < e; ++j) {
+        Node *in = N->getNthInput(j).getNode();
+        if (isa<Storage>(in) || bfs.visited.count(in)) {
+          continue;
+        }
+        nList.push_back(in);
+        bfs.visited.insert(in);
+      }
+    }
+    if (nList.size() > 0) {
+      auto newPair = std::make_pair(level, nList);
+      bfs.levels.push_back(newPair);
+      level++;
+    }
+    current++;
+  }
+
+  return bfs;
+}
+
+/// Assign nodes to partitions and return the mapping.
+NodeToFunctionMap Partitioner::selectPartitions(Function *F,
+                                                unsigned availableMemory) {
+  NodeToFunctionMap mapping;
+  BFSLevel bfs = getBFSLevel(F);
+  unsigned level = bfs.levels.size();
+  // A list of cut. The graph can be partitioned by levels [level - 1,
+  // cut[0]), [cut[0] - 1, cut[1]), ..., [cut[n], -1).
+  std::vector<int> cut;
+
+  // Step 1 : get the initial cut based on BFS levels and avaiableMemory.
+  // TODO .. need to remove the duplicated memory useage.
+  unsigned mem = 0;
+  for (int i = level - 1; i >= 0; i--) {
+    unsigned tmp = 0;
+    for (int j = 0, e = bfs.levels[i].second.size(); j < e; j++) {
+      Node *N = bfs.levels[i].second[j];
+      tmp += memUsage_[N];
+    }
+    if (mem + tmp > availableMemory) {
+      if (mem == 0) {
+        // This means the mem usage for one level exceeds the availableMem,
+        // accept it now and will do adjustment later.
+        cut.push_back(i + 1);
+      } else {
+        cut.push_back(i);
+        mem = tmp;
+      }
+    } else {
+      mem += tmp;
+    }
+  }
+
+  // The last border.
+  cut.push_back(-1);
+
+  // Step 2 : Create the initial mapping between node and functions.
+  for (int k = 0, e = cut.size(); k < e; k++) {
+    auto *newF = F->getParent()->createFunction(std::string(F->getName()) +
+                                                "_part" + std::to_string(k));
+    mapping.createPartition(newF);
+    for (int i = k > 0 ? cut[k - 1] : level - 1; i > cut[k]; i--) {
+      for (int j = 0, e1 = bfs.levels[i].second.size(); j < e1; j++) {
+        Node *N = bfs.levels[i].second[j];
+        mapping.add(N, newF);
+      }
+    }
+  }
+  // Step 3 : adjust the partition based on performance (Advanced Graph
+  // Paritioning algrithm will be applied here).
+  // --- TODO
+
+  return mapping;
+}
+
+/// Adjust the logicalDevice ID for each DAGNode. This happens when \p num (i.e.
+/// the number of DAGNodes) is larger than the number of devices. E.g:
+/// node1(6GB) -> node2(14GB) -> node3(6GB). The memory limitation is 16GB, and
+/// there is only 2 devices.
+void Partitioner::adjustLogicalDeviceID(DAGNode *DAG, int num) {}
```

Comment:
Should we add something like assert(false && "invalid .. .. ");

Reply:
Partitioner will return the minimal k partitions. And the provisioner will double check if k fits the devices and give the invalid assert.